common/src/main/java/com/wynntils/gui/screens/CharacterSelectorScreen.java

```diff
@@ -0,0 +1,471 @@
+                Texture.CHALLENGES_ICON.height());
+        // FIXME: I have no idea what to render here..
+
+        poseStack.popPose();
+    }
+
+    private void renderScrollButton(PoseStack poseStack) {
+        RenderUtils.drawScalingTexturedRect(
+                poseStack,
+                Texture.CHARACTER_SELECTION_SCROLL_BUTTON.resource(),
+                Texture.LIST_BACKGROUND.width() * currentTextureScale * 0.916f,
+                MathUtils.map(
+                        scrollOffset,
+                        0,
+                        classInfoButtons.size() - CHARACTER_INFO_PER_PAGE,
+                        Texture.LIST_BACKGROUND.height() * currentTextureScale * 0.01f,
+                        Texture.LIST_BACKGROUND.height() * currentTextureScale * 0.92f),
+                0,
+                Texture.CHARACTER_SELECTION_SCROLL_BUTTON.width() * currentTextureScale,
+                Texture.CHARACTER_SELECTION_SCROLL_BUTTON.height() * currentTextureScale,
+                Texture.CHARACTER_SELECTION_SCROLL_BUTTON.width(),
+                Texture.CHARACTER_SELECTION_SCROLL_BUTTON.height());
+    }
+
+    private void renderButtons(PoseStack poseStack, int mouseX, int mouseY, float partialTick) {
+        for (Widget renderable : this.renderables) {
+            renderable.render(poseStack, mouseX, mouseY, partialTick);
+        }
+
+        for (int i = scrollOffset; i < Math.min(classInfoButtons.size(), scrollOffset + CHARACTER_INFO_PER_PAGE); i++) {
+            classInfoButtons.get(i).render(poseStack, mouseX, mouseY, partialTick);
+        }
+    }
+
+    private void renderPlayer() {
+        McUtils.player().setInvisible(false);
+        // This is actually needed...
+        McUtils.player().resetFallDistance();
+        McUtils.player().setSwimming(false);
+
+        int scale = this.height / 5;
+        InventoryScreen.renderEntityInInventory(
+                (int) (this.width * 0.6f), (int) (this.height * 0.85f), scale, 0, 0, McUtils.player());
+    }
+
+    private void setScrollOffset(int delta) {
+        scrollOffset = MathUtils.clamp(scrollOffset - delta, 0, classInfoButtons.size() - CHARACTER_INFO_PER_PAGE);
+    }
+
+    private void reloadQuestInfoButtons() {
+        this.selected = null;
+        this.scrollOffset = 0;
+        classInfoButtons.clear();
+
+        final float width = Texture.LIST_BACKGROUND.width() * currentTextureScale * 0.9f - 3;
+        final int height = Math.round(Texture.LIST_BACKGROUND.height() * currentTextureScale / 8f);
+        for (int i = 0; i < classInfoList.size(); i++) {
+            ClassInfo classInfo = classInfoList.get(i);
+            classInfoButtons.add(new ClassInfoButton(
+                    5, (5 + ((i % CHARACTER_INFO_PER_PAGE) * height)), (int) width, height, classInfo, this));
+        }
+    }
+
+    private static ClassInfo getClassInfoFromItem(ItemStack item, int slot, String className) {
+        ClassType classType = null;
```

Comment:
Regarding ClassInfo/CharacterInfo: I think this is very similar to `CharacterInfo.parseCharacter`. Granted, this extracts a bit more, but that should really be done for `parseCharacter` as well. In the future, we will need the XP and finished quests info, to help us confirm that we match to a specific character.

We're ending up with CLASS_ITEM_LEVEL_PATTERN and CLASS_MENU_LEVEL_PATTERN being virtually the same, etc. This is exactly the situation we had in Legacy, where we did not design a generic "model" API which could be used by all parts of the code needing it, so each and every place reinvented it themselves.

Reply:
But then, where would we need the XP or quest info, when those can be updated mid gameplay, and we need to get it from somewhere else otherwise then?

Reply:
I always wanted to have some generic regex class.

Reply:
So in my mind, the CharacterManager would provide us with e.g. "XP" from the best available source. Before selecting a character, that is the lore from the character selection screen. After logging in, it is the vanilla experience level. But the outwards interface is the same, just "getXp()".

Reply:
> I always wanted to have some generic regex class.

That sounds dangerous. :-) Do you mean like a way to help us create regexp:s locally where they belong (if so, I think it's fine but probably a bit futile to get something that is genuinely better than plain regexps), or do you mean that you want to collect a lot of regexps from all over the place in a single file? If you mean the latter, I think you're just creating chaos. Instead, I think the mentality should be: Why do you need a regexp? To extract some "wynncrafty" thing, right? So then encapsulate the result of this regexp in a Model and present it in a "programmer friendly" way.

Reply:
> > I always wanted to have some generic regex class.
> 
> That sounds dangerous. :-) Do you mean like a way to help us create regexp:s locally where they belong (if so, I think it's fine but probably a bit futile to get something that is genuinely better than plain regexps), or do you mean that you want to collect a lot of regexps from all over the place in a single file? If you mean the latter, I think you're just creating chaos. Instead, I think the mentality should be: Why do you need a regexp? To extract some "wynncrafty" thing, right? So then encapsulate the result of this regexp in a Model and present it in a "programmer friendly" way.

You are probably right, it is just an easy way out :)